File: two_pointers/3397.py
class Solution:
    def maxDistinctElements(self, nums: List[int], k: int) -> int:
        n = len(nums)
        nums.sort()
        count = 0
        prev = float("-inf")
        for i in range(n):
            minVal = nums[i] - k
            if prev < minVal:
                prev = minVal
                count += 1
            elif prev < nums[i] + k:
                prev += 1
                count += 1
        return count

File: two_pointers/3397.cpp

// Tc = O(n.logn) + n
// Sc = O(1) [some can say O(n) for sorting]
class Solution {
public:
    int maxDistinctElements(vector<int>& nums, int k) {
        int n = nums.size();
        sort(begin(nums), end(nums));
        int count = 0;
        int prev = INT_MIN;
        for (int i = 0; i < n; i++) {
            int minVal = nums[i] - k;
            if (prev < minVal) {
                prev = minVal;
                count++;
            } else if (prev < nums[i] + k) {
                prev = prev + 1;
                count++;
            }
        }
        return count;
    }
};